### ISSUE TITLE
Generating hashes of large images crashes the server and screws-up the GUI

### ISSUE THREAD

**Issue by WhiteWinterWolf** [GNS3/gns3-gui#2239]
When adding a node of a newly Qemu virtual machine newly added to GNS3, GNS3 generates a *.md5sum* file containing the hash of the virtual machine HDD image.

In case of relatively large disk images (more than a dozen of gigabytes, which is likely for any modern operating system), generating the MD5 sum takes some times (manually generating the sum takes about 4 minutes on my system) but GNS3 immediately enters in a loop raising timeout messages:

    2017-08-27 16:20:21 INFO modules.qemu:177 instantiating node <class 'gns3.modules.qemu.qemu_vm.QemuVM'>
    2017-08-27 16:20:21 INFO modules.qemu.qemu_vm:44 QEMU VM instance is being created
    2017-08-27 16:20:21 INFO modules.qemu:189 creating node QEMU VM with id 2
    2017-08-27 16:20:35 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/computes
    2017-08-27 16:20:35 ERROR compute_manager:76 Error while getting compute list: Operation timeout
    2017-08-27 16:20:50 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/computes
    2017-08-27 16:20:50 ERROR compute_manager:76 Error while getting compute list: Operation timeout
    2017-08-27 16:21:05 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/computes
    2017-08-27 16:21:05 ERROR compute_manager:76 Error while getting compute list: Operation timeout
    2017-08-27 16:21:20 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/computes
    2017-08-27 16:21:20 ERROR compute_manager:76 Error while getting compute list: Operation timeout
    2017-08-27 16:21:35 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/computes
    2017-08-27 16:21:35 ERROR compute_manager:76 Error while getting compute list: Operation timeout
    2017-08-27 16:21:50 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/computes
    2017-08-27 16:21:50 ERROR compute_manager:76 Error while getting compute list: Operation timeout
    2017-08-27 16:22:05 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/computes
    2017-08-27 16:22:05 ERROR compute_manager:76 Error while getting compute list: Operation timeout
    2017-08-27 16:22:20 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/projects/917db82a-7327-41e6-b171-cf9767f367fc/nodes
    2017-08-27 16:22:20 CRITICAL qt:109 Error while setting up node: Operation timeout
    2017-08-27 16:22:20 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/computes
    2017-08-27 16:22:21 ERROR compute_manager:76 Error while getting compute list: Operation timeout
    2017-08-27 16:22:23 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:22:23 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:22:28 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:22:28 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:22:33 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:22:33 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:22:35 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/computes
    2017-08-27 16:22:35 ERROR compute_manager:76 Error while getting compute list: Operation timeout
    2017-08-27 16:22:38 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:22:38 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:22:43 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:22:43 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:22:48 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:22:48 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:22:50 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/computes
    2017-08-27 16:22:50 ERROR compute_manager:76 Error while getting compute list: Operation timeout
    2017-08-27 16:22:53 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:22:53 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:22:58 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:22:59 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:23:03 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:23:03 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:23:05 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/computes
    2017-08-27 16:23:05 ERROR compute_manager:76 Error while getting compute list: Operation timeout
    2017-08-27 16:23:05 WARNING http_client:572 Aborting request for http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:23:05 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:23:05 WARNING http_client:572 Aborting request for http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:23:05 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:23:05 WARNING http_client:572 Aborting request for http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:23:05 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:23:07 WARNING http_client:572 Aborting request for http://admin@127.0.0.1:3080/v2/computes
    2017-08-27 16:23:07 ERROR compute_manager:76 Error while getting compute list: Operation timeout
    2017-08-27 16:23:08 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:23:08 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:23:13 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:23:13 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:23:18 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:23:18 ERROR local_config:153 Can't get settings from controller
    ^C2017-08-27 16:23:22 INFO main:280 Signal received exiting the application
    2017-08-27 16:23:22 WARNING http_client:572 Aborting request for http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:23:22 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:23:22 WARNING http_client:572 Aborting request for http://admin@127.0.0.1:3080/v2/computes
    2017-08-27 16:23:22 ERROR compute_manager:76 Error while getting compute list: Operation timeout
    2017-08-27 16:23:22 WARNING http_client:572 Aborting request for http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:23:22 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:23:22 WARNING http_client:572 Aborting request for http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:23:22 ERROR local_config:153 Can't get settings from controller
    2017-08-27 16:23:22 INFO local_config:423 Section MainWindow has changed. Saving configuration
    2017-08-27 16:23:22 INFO local_config:299 Configuration save to /home/gzbk/.config/GNS3/gns3_gui.conf
    2017-08-27 16:23:22 INFO local_server:542 Stopping local server (PID=1628)
    2017-08-27 16:23:23 WARNING http_client:559 Timeout request http://admin@127.0.0.1:3080/v2/settings
    2017-08-27 16:23:23 ERROR local_config:153 Can't get settings from controller
    ^C2017-08-27 16:23:32 INFO main:280 Signal received exiting the application
    ^C2017-08-27 16:23:33 INFO main:280 Signal received exiting the application
    ^C2017-08-27 16:23:34 INFO main:280 Signal received exiting the application
    ^C2017-08-27 16:23:40 INFO main:280 Signal received exiting the application
    ^Z
    [1]+  Stopped                 LC_ALL=fr_FR.UTF8 gns3
    gzbk@labo:~$ kill -9 %%

    [1]+  Stopped                 LC_ALL=fr_FR.UTF8 gns3
    gzbk@labo:~$ 
    [1]+  Killed                  LC_ALL=fr_FR.UTF8 gns3

When the GUI finally gives up by closing it processing window and popup the *"Error while setting up node: Operation timeout"* error message, the server seems dead and it becomes impossible to normally close the GUI anymore, the same kind of nasty logic than the one I mention in #2237: GNS3 cannot stop the GUI because the server is not started.

`kill -9` is the only way to shut GNS3 down.

**Workaround**

The workaround is to manually create a snapshot file like this

    qemu-img create -f qcow2 -o backing_file=original-file.qcow2 snapshot-file.qcow2

And then provide the snapshot file to GNS3 instead of the large, original HDD image file.

Personally, I don't see the added value of this MD5 sum computation and would gladly see it disappear.

**Comment by WhiteWinterWolf** [comment]
When I say *"I don't see the added value of this MD5 sum computation and would gladly see it disappear"*, I mean:

- I suppose that this files allows GNS3 to ensure that the image file hasn't been modified outside of GNS3, however I can see use-cases where such arbitrary limitation would be too restrictive (for instance to allow to boot a non-link based VM outside of GNS3 to apply system updates).

- Such kind of test is ineffective: see the workaround, the checksum will only detect a modification of the snapshot file and not of its backing file. One could imagine GNS3 to actually parse all types of multi-files HDD images (Qemu chained backing files, VMware sparse files, etc.) and generate a hash for each involved file, but this would be complex and still unreliable (what if a users commits a snaphot? File content change but the VM content remains the same so GNS3 should not reject it. What if the base file is stored on a shared read-only location? GNS3 won't be able to create its *.md5sum* file).

While I see the multiple issues that such feature raises, I cannot see any real advantage of it, hence my final note that would it be removed I won't miss it.

**Comment by noplay** [comment]
The md5 is here to allow you to reload your project and be sure your are using the correct images. It's for project sharing wihout providing the images to others.

But the generation should not be an issue because at startup we generate the .md5sum and when you upload an image also. 

How your images are stored? A network share? An HDD (5400 rpm) ? An SSD (7200 rpm) ?

**Comment by WhiteWinterWolf** [comment]
Hi,
I'm using a 2.5 inch 5400 rpm hard-disk.

Ideally, I would like the MD5 sum calculation to only affect use cases where it is really needed. 
For instance, if it is used when exporting or importing a project, I would like the MD5 sum to be calculated as part of the export / import process and not in advance *in case if maybe at some point later I may* be interested in sharing the project with other people.

But, that being said, I am not familiar with GNS3 project sharing features so it may go well beyond the export / import functionality and these MD5 may be required for other people (but even then, what about a setting in GNS3 allowing to disable the automatic generation of these files?).

In all cases, the current behavior remains wrong. Calculating the MD5 sum of a 20GB image on my system takes about 4 minutes:

```
user@host:~$ ls -lh hda.qcow2 
-rw-r--r-- 1 user user 23G Mar 21  2017 hda.qcow2
user@host:~$ time md5sum hda.qcow2 
8358f748c93fd3e8ad8e6ea3d82ab392  hda.qcow2

real	4m4.814s
user	1m26.532s
sys	0m14.960s
user@host:~$
```

According to the logs provided in the original post, GNS3 times-out and screws itself up after 15 seconds.

If these MD5 files are so required, then IMHO GNS3 should react one of the following ways:

- GNS3 must refuse to load disk image weighting more than a few hundreds of MB or raise a clean and proper error message if the hash generation takes more that 15 seconds (with no crash of the server and the GUI remaining usable after the error). Larger images being common, in case of such error GNS3 should tell the user the recommended way to still be able to load his image in GNS3 (see my workaround in the original post for instance, it however only addresses Qemu images).

- GNS3 should wait the time required to generate the hash, even if this means waiting about an hour to generate the hash of an image of a few hundred gigabytes.
  The user should:
    - Get a feedback about the currently ongoing task (ideally with a progression bar, but at least a message telling something like "Calculating disk image MD5 sum, this may take some time...").
    - Have a *Cancel* button allowing him to cleanly abort the task if it takes too long.

**Comment by noplay** [comment]
You can also just share .gns3 it's very common. Or restore it from a backup.

How do you import the file ?  Via the GUI or cp in the images folder?

**Comment by WhiteWinterWolf** [comment]
Hi,

I do everything through the GUI, as follow:

1. Create a new Qemu VM end-device template, select an existing image file as disk image, don't copy it into GNS3 directory tree.
2. Create a new topology, and add this new end-device in the topology, this is when the issue appears.

After that, the only way to stop the GNS3 GUI is to brutally kill it (`kill -9`), and the only way to use the new end-device is by using a snapshot file as intermediary between GNS3 and the actual disk image as described in the workaround.

**Comment by WhiteWinterWolf** [comment]
Other effects I attribute to the same issue:

- As soon as a topology involves a large disk images, trying to do a snapshot generates the timeouts and general wreak havoc of the GNS3 mostly resulting in having to brutally kill it and restart the host (if at least GNS3 would be able to cleanly stop itself!). This makes renders the snapshot feature unusable in such conditions.

- It is not possible to link a large DVD image (system installation media) to an end-node: this also breaks everything.

I don't have workaround for the above mentioned issue, just considering these features as broken and moving-on.

I sincerely hope that this will be fixed soon as I find quite frustrating to see an otherwise good and useful tool to fail so hard on such a silly thing.

**Comment by cioby23** [comment]
I think maybe md5 hashes should apply only to images installed from an appliance file. There is no point to generate md5 hashes for an image which the user creates himself.

**Comment by ziajka** [comment]
Above-mentioned issue is connected with this part of code:

https://github.com/GNS3/gns3-server/blob/2.1/gns3server/controller/node.py#L328

Removing limitation of 120 seconds may be risky. What do you think?

**Comment by grossmj** [comment]
Indeed that looks risky.

**Comment by ziajka** [comment]
Increased timeout, needs further changes in a way we do that.

**Comment by cdwchriburg** [comment]
It seems the issue with the MD5 generation is multi-fold
 - Performance is CPU-bound due to single-threaded nature
 - Due to being single-threaded, large files take a long time to calculate the MD5 even on very powerful systems
 - The code performing this MD5 generation appears to be blocking in some way that prevents other GUI clients from getting responses from GNS3-Server process

Fixing root cause on this would seem to require spinning off the MD5 generation to its own thread and avoid blocking whatever thread/process is handling the other GNS3-Server functions.

**Comment by ziajka** [comment]
Ref. https://github.com/GNS3/gns3-server/pull/1275

**Comment by ziajka** [comment]
Hi @WhiteWinterWolf,

We've changed MD5 sums calculations to be on thread, can you check our 2.2 dev branch? (Build from sources or Nightly Builds).
Please be always careful with dev versions as it can be dangerous to use them and it should be done in separate environment. 

Thanks,
Dominik

**Comment by WhiteWinterWolf** [comment]
Hi @ziajka,

Thank you for the notification. I will be away for a week but will surely test this ASAP.
Thank you a lot for your work on this!

Regards,
Simon.

**Comment by ziajka** [comment]
Thanks @WhiteWinterWolf!

**Comment by WhiteWinterWolf** [comment]
Hi @ziajka,

As for reminder, on my main system I'm (still) currently using 2.0.3.

In a VM, I first tried 2.1.3, two remarks:

1/ While creating the new Qemu VM in GNS3, there is an odd delay (several minutes for a 20G large image) after having selected the file image path and before GNS3 asks me if I want to copy it to the default image directory (no progress bar, log entry or message box telling me what is happening, the GUI just seem frozen while the HDD is scratching). GNS3 seems to process something with the image file, but what I don't know (no MD5 sum file generated, apparently no file copied yet). Maybe there is a new integrity check of some sort, just guessing. In 2.0.3 there is no such delay. Anyway this processing eventually reach to a end, so this is not a big deal.

2/ The really blocking issue I raised in this ticket seem to be already solved in 2.1.3 : the timeout messages regarding *http://admin@127.0.0.1:3080/v2/computes* are still present, but the GUI is now able to restore back its functionalities once the MD5 processing is finished (I had no more timeout messages regarding *http://admin@127.0.0.1:3080/v2/settings*, but this may have been a side-effect of the MD5 calculation being faster thanks to the VM image being already present in the system's cache). So, as when selecting the image file path, it's just a matter of waiting the first time the VM is used in GNS3, so again not a real problem (coffee has been invented precisely to handle such waiting time ;) !).

Regarding 2.2dev, I'm a bit sceptical. The template VM creation goes exactly as with 2.1.3, but when trying to add the newly created template VM to a topology this failed with the following error message:

`2018-03-11 18:55:40 ERROR appliance_manager:107 Error while creating node: /home/user/hdd.qcow2 is not allowed on this remote server. Please use only a filename in /home/user/GNS3/images/QEMU.`

I did not understood at first. It seems to be a new requirement to store the all the image files below the */GNS3/images/QEMU* directory. I added my homedir to the *Additional search locations for binary images* but with no success. I ended-up manually moving the image file to the requested path.

I got the *http://admin@127.0.0.1:3080/v2/settings* messages back and the MD5 took a very long time (a bit more than 10 minutes) to be computed, but I guess when using 2.1.3 the image file was still present in the system's cache (a reboot occurred in between) which could explain the long processing time (I did not optimize the HDD I/O for this VM, so 10 minutes compared to the 4 minutes I noticed on the physical host in my initial message doesn't seem unreasonable to me).

Nevertheless, at the end, once the MD5 calculation ended the GUI was still able to get back on its feet and remain usable, which is all I need (the dozens of warning and error messages generated in between are arguably messy, but as long as it works, I suppose that handling large image files is a narrow use-case so I can understand such behavior).

For me, since GNS3 server and GUI can now able to handle large VM image file without breaking, this answers my need so unless you want to improve some other aspects this ticket can be closed.

Thank you all for your work on this!

**Comment by ziajka** [comment]
Thank you @WhiteWinterWolf for really great feedback. This won't be our last word about improving the way of uploading/managing images. For now we we're focus on Web UI and later we will go back to that.
